src/sql/src/plan/query.rs

```diff
@@ -1207,56 +1250,72 @@ fn check_col_index(name: &str, e: &Expr<Aug>, max: usize) -> Result<Option<usize
     }
 }
 
-fn plan_query(
-    qcx: &mut QueryContext,
-    q: &Query<Aug>,
-) -> Result<(HirRelationExpr, Scope, RowSetFinishing, GroupSizeHints), PlanError> {
+struct PlannedQuery {
+    expr: HirRelationExpr,
+    scope: Scope,
+    order_by: Vec<ColumnOrder>,
+    limit: Option<HirScalarExpr>,
+    offset: usize,
+    project: Vec<usize>,
+    group_size_hints: GroupSizeHints,
+}
+
+fn plan_query(qcx: &mut QueryContext, q: &Query<Aug>) -> Result<PlannedQuery, PlanError> {
     qcx.checked_recur_mut(|qcx| plan_query_inner(qcx, q))
 }
 
-fn plan_query_inner(
-    qcx: &mut QueryContext,
-    q: &Query<Aug>,
-) -> Result<(HirRelationExpr, Scope, RowSetFinishing, GroupSizeHints), PlanError> {
+fn plan_query_inner(qcx: &mut QueryContext, q: &Query<Aug>) -> Result<PlannedQuery, PlanError> {
     // Plan CTEs and introduce bindings to `qcx.ctes`. Returns shadowed bindings
     // for the identifiers, so that they can be re-installed before returning.
     let cte_bindings = plan_ctes(qcx, q)?;
 
     let limit = match &q.limit {
         None => None,
         Some(Limit {
-            quantity: Expr::Value(Value::Number(x)),
+            quantity,
             with_ties: false,
-        }) => Some(x.parse()?),
+        }) => {
+            let ecx = &ExprContext {
+                qcx,
+                name: "LIMIT",
+                scope: &Scope::empty(),
+                relation_type: &RelationType::empty(),
+                allow_aggregates: false,
+                allow_subqueries: true,
```

Comment:
@benesh we might want to disallow subqueries here as well. At least for LIMITs parsed in the top query this results in [the error reported here](https://github.com/MaterializeInc/materialize/pull/23276#issuecomment-1845843641).